### PR TITLE
Deactivate repos the user can't see anymore

### PR DESCRIPTION
**Pull request by thorncp**
If the user cannot see a repo and the repo has no other memberships. It
will be inactivated and the memberships will be destroyed.

https://trello.com/c/SBKDsHjl

### CODE REVIEW

File: app/services/repo_synchronization.rb [hunk 1]
```diff
@@ -9,15 +9,47 @@ def api
   end
 
   def start
-    user.repos.clear
+    update_user_repos_from_api
+    deactivate_invisble_repos
```
Comment (malandrina) [inline app/services/repo_synchronization.rb]: Typo: "invisible"

File: spec/services/repo_synchronization_spec.rb [hunk 2]
```diff
@@ -118,5 +93,70 @@
         expect(second_user.reload.repos.size).to eq(1)
       end
     end
+
+    describe "deleted repos" do
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: Could we make this more descriptive? `describe "when repo has been deleted"`, maybe?

File: spec/services/repo_synchronization_spec.rb [hunk 3]
```diff
@@ -118,5 +93,70 @@
...
+      it "inactivates a repo when there are no other memberships" do
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: Prefer "deactivate" to be consistent with the rest of the code.

File: spec/services/repo_synchronization_spec.rb [hunk 4]
```diff
@@ -118,5 +93,70 @@
...
+        expect(repo).to be_active
+      end
+
+      it "destroys the users membership when a repo gets inactivated" do
+        repo = create(:repo, :active)
+        repo.memberships.destroy_all
+        membership = create(:membership, repo: repo)
+        user = membership.user
+        github_token = "githubtoken"
+
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: Style nitpick: Looks like lines 151-153 are part of the setup phase, so this newline can be removed.

File: spec/services/repo_synchronization_spec.rb [hunk 5]
```diff
@@ -118,5 +93,70 @@
...
+        expect(membership.user.repos).to include(second_repo)
+      end
+
+      it "does not inactivate a repo when there are other memberships" do
+        first_membership = create(:membership)
+        repo = first_membership.repo
+        repo.update(active: true)
+        membership = create(:membership, repo: repo)
+        github_token = "githubtoken"
+
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: See below comment about removing the newline.

File: spec/services/repo_synchronization_spec.rb [hunk 6]
```diff
@@ -118,5 +93,70 @@
         expect(second_user.reload.repos.size).to eq(1)
       end
     end
+
+    describe "deleted repos" do
+      it "inactivates a repo when there are no other memberships" do
+        repo = create(:repo, :active)
+        repo.memberships.destroy_all
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: Is this necessary?

File: spec/services/repo_synchronization_spec.rb [hunk 7]
```diff
@@ -118,5 +93,70 @@
...
+          full_name: second_repo.full_github_name,
+          id: second_repo.github_id,
+          private: true,
+          owner: {
+            type: "User"
+          }
+        }
+        resource = double(:resource, to_hash: attributes)
+        github_token = "githubtoken"
+        api = double(:github_api, repos: [resource])
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: Consider extracting lines 114-115 to a private method that can be called here and on lines 134-135 (and 151-152).

File: spec/services/repo_synchronization_spec.rb (hunk 2 above)
Comment (thorncp) [inline spec/services/repo_synchronization_spec.rb]: This should probably be something like `"when a user no longer has access to a repo"`, because we don't know that the repository was deleted, only that the user can no longer see it. Jake and I tried out several terms/phrases for this, and settled on `invisible` in one of the method names. I don't know if that's the most clear way to say that though. Any thoughts?

File: app/services/repo_synchronization.rb [hunk 8]
```diff
@@ -9,15 +9,47 @@ def api
...
     end
   end
 
-  private
+  def deactivate_repo(repo)
+    repo.deactivate
+    repo.memberships.destroy_all
+  end
+
+  def update_user_repos_from_api
```
Comment (malandrina) [inline app/services/repo_synchronization.rb]: Looks like there's a lot going on here. What do you think about moving the memoization of repos returned by the API to a separate method? Something like this:

```
def api_repos
  @api_repos ||= api.repos.map do |resource|
    find_or_create_repo(resource)
  end
end

def update_user_repos_from_api
  api_repos.each do |repo|
    unless user_repos.include?(repo)
      user.repos << repo
    end
  end
end
```

File: spec/services/repo_synchronization_spec.rb (hunk 6 above)
Comment (thorncp) [inline spec/services/repo_synchronization_spec.rb]: We added that because the [after create hook](https://github.com/thoughtbot/hound/blob/1c86a7bb55907ce522a769bc54dfe309fc3c0c30/spec/factories.rb#L24-L28) in the factory was causing us to have more memberships than we wanted in this spec. It's definitely a bandaid, but the only way we thought of to fix that was to remove the hook completely. We didn't want to do that in this PR though, since it would likely require updating other specs.

Is there a more sensible approach? I feel like the factories shouldn't have any hooks at all. It took us awhile to track down where the extra records were being creating, which was pretty frustrating.

File: spec/services/repo_synchronization_spec.rb (hunk 2 above)
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: `"when a user no longer has access to a repo"` sounds good to me!

File: spec/services/repo_synchronization_spec.rb (hunk 6 above)
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: Got it. In that case I think it's fine. We could move the hook to a trait (`:with_user` ?), but that's a refactoring and should be done in a separate PR.

File: app/services/repo_synchronization.rb (hunk 8 above)
Comment (thorncp) [inline app/services/repo_synchronization.rb]: I like it.

File: spec/services/repo_synchronization_spec.rb (hunk 6 above)
Comment (thorncp) [inline spec/services/repo_synchronization_spec.rb]: :+1: for the trait

**Comment by thorncp** [comment]
Needs re-review

File: spec/services/repo_synchronization_spec.rb [hunk 9]
```diff
@@ -118,5 +88,71 @@
...
+        stub_api_repos(repos: [])
+        synchronization = RepoSynchronization.new(membership.user, github_token)
+
+        synchronization.start
+        repo.reload
+
+        expect(repo).to be_active
+      end
+
+      it "destroys the users membership when a repo gets inactivated" do
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: "inactivated" => "deactivated"

File: spec/services/repo_synchronization_spec.rb [hunk 10]
```diff
@@ -118,5 +88,71 @@
...
+        synchronization.start
+        repo.reload
+        second_repo.reload
+
+        expect(repo).not_to be_active
+        expect(second_repo).to be_active
+        expect(membership.user.repos).to include(second_repo)
+      end
+
+      it "does not inactivate a repo when there are other memberships" do
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: "inactivate" => "deactivate"

File: spec/services/repo_synchronization_spec.rb (hunk 9 above)
Comment (thorncp) [inline spec/services/repo_synchronization_spec.rb]: Apparently my grep needs work :confounded:

File: spec/services/repo_synchronization_spec.rb [hunk 11]
```diff
@@ -118,5 +88,71 @@
...
+        stub_api_repos(repos: [resource])
+
+        synchronization = RepoSynchronization.new(membership.user, github_token)
+
+        synchronization.start
+        repo.reload
+        second_repo.reload
+
+        expect(repo).not_to be_active
+        expect(second_repo).to be_active
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: I'm not sure that lines 118-119 are useful. What is the value of these assertions?

File: spec/services/repo_synchronization_spec.rb [hunk 12]
```diff
@@ -118,5 +88,71 @@
...
+        expect(membership.user.repos).to include(second_repo)
+      end
+
+      it "does not deactivate a repo when there are other memberships" do
+        first_membership = create(:membership)
+        repo = first_membership.repo
```
Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: Lines 123-124 could be condensed to `repo = create(:membership).repo`.

File: app/services/repo_synchronization.rb [hunk 13]
```diff
@@ -9,15 +9,51 @@ def api
   end
 
   def start
-    user.repos.clear
+    update_user_repos_from_api
+    deactivate_invisible_repos
```
Comment (malandrina) [inline app/services/repo_synchronization.rb]: Here's a thought: I'm not sure the word "invisible" adds much here because it is kind of ambiguous, so this could become simply `deactivate_repos`. Thoughts?

File: spec/services/repo_synchronization_spec.rb (hunk 11 above)
Comment (thorncp) [inline spec/services/repo_synchronization_spec.rb]: I think we added these because our implementation was deactivating *all* repos. However, this should be covered by other specs, and I think it was only meant to be temporary. I'll dig into it further.

File: app/services/repo_synchronization.rb (hunk 13 above)
Comment (thorncp) [inline app/services/repo_synchronization.rb]: I think that would be fine. I believe our thought process was something like "but *what* repos would be deactivated?". Perhaps we simply overthought it, given that it's a private method. I agree that "invisible" really has no meaning here.

File: spec/services/repo_synchronization_spec.rb (hunk 11 above)
Comment (thorncp) [inline spec/services/repo_synchronization_spec.rb]: Yep, totally superfluous.

Comment (malandrina) [inline spec/services/repo_synchronization_spec.rb]: :cool:

**Comment by malandrina** [comment]
Might want to get another set of eyes on this to validate the business logic, but it looks good to me.

File: app/services/repo_synchronization.rb [hunk 14]
```diff
@@ -9,15 +9,51 @@ def api
...
+      if repo_not_visible_to_anyone?(repo)
+        deactivate_repo(repo)
+      end
     end
   end
 
-  private
+  def deactivate_repo(repo)
+    repo.deactivate
+    repo.memberships.destroy_all
```
Comment (gylaz) [inline app/services/repo_synchronization.rb]: Should we remove subscription in this case too?

File: app/services/repo_synchronization.rb [hunk 15]
```diff
@@ -9,15 +9,51 @@ def api
...
+  def update_user_repos_from_api
+    api_repos.each do |repo|
+      unless user_repos.include?(repo)
+        user.repos << repo
+      end
+    end
+  end
+
+  def repo_not_visible_to_anyone?(repo)
+    !api_repos.include?(repo) && repo.memberships.count == 1
```
Comment (gylaz) [inline app/services/repo_synchronization.rb]: If the repo is not part of this user's repos, but some other user does have a membership to it, this logic would be flawed...?

File: app/services/repo_synchronization.rb (hunk 14 above)
Comment (thorncp) [inline app/services/repo_synchronization.rb]: Yes, yes we should.

File: spec/services/repo_synchronization_spec.rb [hunk 16]
```diff
@@ -118,5 +88,55 @@
         expect(second_user.reload.repos.size).to eq(1)
       end
     end
+
+    describe "when a user no longer has access to a repo" do
+      it "deactivates a repo when there are no other memberships" do
+        repo = create(:repo, :active)
+        repo.memberships.destroy_all
+        membership = create(:membership, repo: repo)
```
Comment (gylaz) [inline spec/services/repo_synchronization_spec.rb]: Instead of lines 94-95 what do you think about grabbing the repo from `memberships`?

File: app/services/repo_synchronization.rb (hunk 15 above)
Comment (thorncp) [inline app/services/repo_synchronization.rb]: This method is called inside a loop over the user's repos, so I don't think that will ever be the case.

File: spec/services/repo_synchronization_spec.rb (hunk 16 above)
Comment (thorncp) [inline spec/services/repo_synchronization_spec.rb]: The default repo factory sets active to false. So, if we use the membership's repo, we'd have to add a line to set that repo to active. Unless there's some way to create a membership and pass the `:active` trait through to the repo? Even there isn't, updating the repo's active flag would probably more clear than what's here.

File: app/services/repo_synchronization.rb [hunk 17]
```diff
@@ -9,15 +9,51 @@ def api
...
   def start
-    user.repos.clear
+    update_user_repos_from_api
+    deactivate_repos
+  end
+
+  private
+
+  def user_repos
+    @user_repos ||= user.repos
```
Comment (gylaz) [inline app/services/repo_synchronization.rb]: I don't think this memoization is necessary. `user.repos` will cache the db call.

File: app/services/repo_synchronization.rb (hunk 15 above)
Comment (gylaz) [inline app/services/repo_synchronization.rb]: Gotcha, the `repo.memberships.count == 1` assumes only the current user's membership. Makes sense now.

**Comment by gylaz** [comment]
Looks good to me.

File: app/services/repo_synchronization.rb (hunk 15 above)
Comment (thorncp) [inline app/services/repo_synchronization.rb]: Yep. I'm definitely open to feedback for making this more clear, though.

**Comment by thorncp** [comment]
Updated to kill the subscription if necessary.

**Comment by gylaz** [comment]
Once tested on staging, ready to be merged.

File: app/services/repo_synchronization.rb [hunk 18]
```diff
@@ -9,15 +9,54 @@ def api
...
+  private
+
+  def user_repos
+    user.repos
+  end
 
-    api.repos.each do |resource|
-      attributes = repo_attributes(resource.to_hash)
-      user.repos << Repo.find_or_create_with(attributes)
+  def deactivate_repos
```
Comment (salbertson) [inline app/services/repo_synchronization.rb]: Would `deactivate_orphaned_repos` make more sense?

**Comment by salbertson** [comment]
@thorncp, is it possible to have a repo in our database without any memberships that hasn't been deleted on GitHub?

I'm wondering if we really want to disable Hound and unsubscribe, if there's a subscription, whenever there are no memberships.

**Comment by thorncp** [comment]
@salbertson that's possible if the user who enabled the repo was the only person with access to it that has logged into hound, and then they are removed from that repo on github. We certainly don't have to disable and unsubscribe if a subscription exists, but that seems like an aggressive approach to me.

File: app/services/repo_synchronization.rb (hunk 18 above)
Comment (thorncp) [inline app/services/repo_synchronization.rb]: That's more clear. I like it.

**Comment by thorncp** [comment]
I discovered a case where memberships weren't being destroyed when the user can no longer see a repo, and other memberships still exist. Fixed in  0814da6

**Comment by thorncp** [comment]
Per Trello discussion, closing this until a real world scenario occurs that forces a solution.
